SCAutolib/controller.py

```diff
@@ -0,0 +1,375 @@
+    local_ca: CA.LocalCA = None
+    ipa_ca: CA.IPAServerCA = None
+    users: [user.User] = None
+
+    @property
+    def conf_path(self):
+        return self._lib_conf_path
+
+    def __init__(self, config: Union[Path, str], params: {} = None):
+        """
+        Constructor will parse and check input configuration file. If some
+        required fields in the configuration are missing, CLI parameters
+        would be checked if missing values are there. If not, an exception
+        would be raised. After parsing the configuration file and filling
+        internal values of the Controller object, other related objects (users,
+        cards, CAs, Authselect, etc.) would be initialised, but any real action
+        that would affect the system
+        wouldn't be made.
+
+        :param config: Path to configuration file with metadata for testing.
+        :type config: pathlib.Path or str
+        :param params: Parameters from CLI
+        :type params: dict
+        :return:
+        """
+
+        # Check params
+
+        # Parse config file
+        self._lib_conf_path = config.absolute() if isinstance(config, Path) \
+            else Path(config).absolute()
+
+        with self._lib_conf_path.open("r") as f:
+            self.lib_conf = json.load(f)
+            assert self.lib_conf, "Data are not loaded correctly."
+        self.lib_conf = self._validate_configuration(params)
+
+    def prepare(self):
+        """
+        Method for setting up whole system based on configuration file and
+        CLI commands
+
+        :return:
+        """
+        ...
+
+    def setup_system(self, install_missing: bool, gdm: bool):
+        """
+        Do general system setup meaning package installation based on
+        specifications in the configuration file, SSSD configuration,
+        configurations for virtual smart cards, etc.
+
+        :param install_missing: If True, all missing packages would be
+            installed
+        :type install_missing:
+        :param gdm: If True, GDM package would be installed
+        :type gdm: bool
+        :return:
+        """
+        LIB_DIR.mkdir(exist_ok=True)
+        LIB_BACKUP.mkdir(exist_ok=True)
+
+        packages = ["opensc", "httpd", "sssd", "sssd-tools", "gnutls-utils"]
+        if gdm:
+            packages.append("gdm")
+
+        # Prepare for virtual cards
+        if "virtual" in [u["card_type"] for u in self.lib_conf["users"]]:
+            packages += ["pcsc-lite-ccid", "pcsc-lite", "virt_cacard",
+                "vpcd", "softhsm"]
+            run("dnf -y copr enable jjelen/vsmartcard")
+
+        # Add IPA packages if needed
+        if not all([u["local"] for u in self.lib_conf["users"]]):
+            packages += self._general_steps_for_ipa()
+
+        # Check for installed packages
+        missing = _check_packages(packages)
+        if install_missing:
+            _install_packages(missing)
+
+        run(['dnf', 'groupinstall', "Smart Card Support", '-y'])
```

Comment:
We may want conditional execution here as Smart Card Support may already be installed at some systems.